e2e/nomostest/testwatcher/watch.go

```diff
@@ -52,37 +56,67 @@ func WatchTimeout(timeout time.Duration) WatchOption {
 	}
 }
 
+// Watcher is used for performing various Watch operations
+type Watcher struct {
+	ctx                context.Context
+	defaultWaitTimeout time.Duration
+	logger             *testlogger.TestLogger
+	restConfig         *rest.Config
+	kubeClient         *testkubeclient.KubeClient
+	scheme             *runtime.Scheme
+}
+
+// NewWatcher constructs a new Watcher
+func NewWatcher(
```

Comment:
When ALL of the fields need to be populated, I usually just skip the constructor and make them all public. That's what I did for the other test package structs. I'm not sure there's really any value in hiding them as private fields. Yes, it makes them not mutable by other packages, but it doesn't really hide any complexity. And it makes a constructor with too many args that you have to then document in the constructor, instead of just on the fields themselves.

Reply:
Also, in the testing utility context, we don't really need to try to protect the user from shooting themselves in the foot as much as we might in a public facing interface that we need to maintain reverse compatibility on.

Reply:
In my view the constructor w/ private fields adds a few things:
- guarantee of immutability from other packages (as you mentioned)
- guarantee of required fields being provided. When looking at the constructor you know that all of the necessary fields have been set for a given instantiation. Conversely, a new person looking at the code can quickly understand what fields they need to provide to create a new instance. If it's just a struct with a bunch of fields, it requires more investigation for a new person to figure out how to instantiate the struct and what fields are required (which may be all or a subset of fields).

I acknowledge that we don't need to be as strict with test utilities compared to a public interface, however I think that mentality is what leads to the situation with `nomostest` where it's an amalgamation of organically grown interfaces. The looseness of the APIs makes it take longer to understand how everything works (i.e. more cyclomatic complexity), rather than easier to use - IMO.